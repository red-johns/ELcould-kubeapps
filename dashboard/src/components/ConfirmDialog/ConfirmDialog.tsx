import * as React from "react";
import * as Modal from "react-modal";
import LoadingSpinner from "../LoadingSpinner";
import "./ConfirmDialog.css";

interface IConfirmDialogProps {
  modalIsOpen: boolean;
  loading: boolean;
  extraElem?: JSX.Element;
  onConfirm: () => Promise<any>;
  closeModal: () => Promise<any>;
}

interface IConfirmDialogState {
  error?: string;
  modalIsOpen: boolean;
}

class ConfirmDialog extends React.Component<IConfirmDialogProps, IConfirmDialogState> {
  public state: IConfirmDialogState = {
    error: undefined,
    modalIsOpen: this.props.modalIsOpen,
  };

  public render() {
    return (
      <div className="ConfirmDialog">
        <Modal
          style={{
            content: {
              bottom: "auto",
              left: "50%",
              marginRight: "-50%",
              right: "auto",
              top: "50%",
              transform: "translate(-50%, -50%)",
            },
          }}
          isOpen={this.props.modalIsOpen}
          onRequestClose={this.closeModal}
          contentLabel="Modal"
        >
          {this.state.error && (
            <div className="padding-big margin-b-big bg-action">{this.state.error}</div>
          )}
          {this.props.loading === true ? (
            <div className="row confirm-dialog-loading-info">
              <div className="col-8 loading-legend">Loading, please wait</div>
              <div className="col-4">
                <LoadingSpinner />
              </div>
            </div>
          ) : (
            <div>
              <div className="margin-b-normal"> Are you sure you want to delete this? </div>
              {this.props.extraElem}
              <button className="button" onClick={this.props.closeModal}>
                Cancel
              </button>
              <button
                className="button button-primary button-danger"
                type="submit"
                onClick={this.props.onConfirm}
              >
                Delete
              </button>
            </div>
          )}
        </Modal>
      </div>
    );
  }

  public openModel = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  public closeModal = () => {
    this.setState({
      modalIsOpen: false,
    });
  };
}

export default ConfirmDialog;
