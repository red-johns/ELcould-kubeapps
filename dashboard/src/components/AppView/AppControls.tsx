import * as React from "react";
import { Redirect } from "react-router";

import { hapi } from "../../shared/hapi/release";
import ConfirmDialog from "../ConfirmDialog";
import LoadingWrapper from "../LoadingWrapper";

interface IAppControlsProps {
  app: hapi.release.Release;
  deleteApp: (purge: boolean) => Promise<boolean>;
}

interface IAppControlsState {
  migrate: boolean;
  modalIsOpen: boolean;
  redirectToAppList: boolean;
  upgrade: boolean;
  deleting: boolean;
  purge: boolean;
}

class AppControls extends React.Component<IAppControlsProps, IAppControlsState> {
  public state: IAppControlsState = {
    deleting: false,
    migrate: false,
    modalIsOpen: false,
    purge: false,
    redirectToAppList: false,
    upgrade: false,
  };

  public render() {
    const { name, namespace } = this.props.app;
    const deleted = this.props.app.info && this.props.app.info.deleted;
    if (!name || !namespace) {
      return <LoadingWrapper />;
    }
    return (
      <div className="AppControls">
        {/* If the app has been deleted hide the upgrade button */}
        {!deleted && (
          <button className="button" onClick={this.handleUpgradeClick}>
            Upgrade
          </button>
        )}
        {this.state.upgrade && (
          <Redirect push={true} to={`/apps/ns/${namespace}/upgrade/${name}`} />
        )}
        <button className="button button-danger" onClick={this.openModel}>
          {deleted ? "Purge" : "Delete"}
        </button>
        <ConfirmDialog
          onConfirm={this.handleDeleteClick}
          modalIsOpen={this.state.modalIsOpen}
          loading={this.state.deleting}
          closeModal={this.closeModal}
          extraElem={
            deleted ? (
              undefined
            ) : (
              <div className="margin-b-normal text-c">
                <label className="checkbox margin-r-big">
                  <input type="checkbox" onChange={this.togglePurge} />
                  <span>Purge release</span>
                </label>
              </div>
            )
          }
        />
        {this.state.redirectToAppList && <Redirect to={`/apps/ns/${namespace}`} />}
      </div>
    );
  }

  public handleUpgradeClick = () => {
    this.setState({ upgrade: true });
  };

  public openModel = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  public closeModal = async () => {
    this.setState({
      modalIsOpen: false,
    });
  };

  public handleDeleteClick = async () => {
    this.setState({ deleting: true });
    // Purge the release if the application has been already deleted
    const alreadyDeleted = this.props.app.info && !!this.props.app.info.deleted;
    const deleted = await this.props.deleteApp(alreadyDeleted || this.state.purge);
    const s: Partial<IAppControlsState> = { modalIsOpen: false };
    if (deleted) {
      s.redirectToAppList = true;
    }
    this.setState(s as IAppControlsState);
  };

  private togglePurge = () => {
    this.setState({ purge: !this.state.purge });
  };
}

export default AppControls;
