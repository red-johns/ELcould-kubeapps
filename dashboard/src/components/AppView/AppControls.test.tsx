import { mount, shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";
import * as ReactModal from "react-modal";
import { Redirect } from "react-router";
import { hapi } from "../../shared/hapi/release";
import itBehavesLike from "../../shared/specs";
import ConfirmDialog from "../ConfirmDialog";

import AppControls from "./AppControls";

it("renders a redirect when clicking upgrade", () => {
  const name = "foo";
  const namespace = "bar";
  const app = new hapi.release.Release({ name, namespace });
  const wrapper = shallow(<AppControls app={app} deleteApp={jest.fn()} />);
  const button = wrapper
    .find(".AppControls")
    .children()
    .find(".button")
    .filterWhere(i => i.text() === "Upgrade");
  expect(button.exists()).toBe(true);
  expect(wrapper.find(Redirect).exists()).toBe(false);

  button.simulate("click");
  const redirect = wrapper.find(Redirect);
  expect(redirect.exists()).toBe(true);
  expect(redirect.props()).toMatchObject({
    push: true,
    to: `/apps/ns/${namespace}/upgrade/${name}`,
  });
});

it("calls delete function when clicking the button", done => {
  const name = "foo";
  const namespace = "bar";
  const app = new hapi.release.Release({ name, namespace });
  const wrapper = shallow(<AppControls app={app} deleteApp={jest.fn(() => true)} />);
  const button = wrapper
    .find(".AppControls")
    .children()
    .find(".button-danger");
  expect(button.exists()).toBe(true);
  expect(button.text()).toBe("Delete");
  button.simulate("click");

  const confirm = wrapper
    .find(".AppControls")
    .children()
    .find(ConfirmDialog);
  expect(confirm.exists()).toBe(true);
  confirm.props().onConfirm(); // Simulate confirmation

  expect(wrapper.state("deleting")).toBe(true);
  // Wait for the async action to finish
  setTimeout(() => {
    wrapper.update();
    const redirect = wrapper
      .find(".AppControls")
      .children()
      .find(Redirect);
    expect(redirect.exists()).toBe(true);
    expect(redirect.props()).toMatchObject({
      push: false,
      to: `/apps/ns/${namespace}`,
    });
    done();
  }, 1);
});

it("calls delete function with additional purge", () => {
  const name = "foo";
  const namespace = "bar";
  const app = new hapi.release.Release({ name, namespace });
  const deleteApp = jest.fn(() => false); // Return "false" to avoid redirect when mounting
  // mount() is necessary to render the Modal
  const wrapper = mount(<AppControls app={app} deleteApp={deleteApp} />);
  ReactModal.setAppElement(document.createElement("div"));
  wrapper.setState({ modalIsOpen: true });
  wrapper.update();

  // Check that the checkbox changes the AppControls state
  const confirm = wrapper.find(ConfirmDialog);
  expect(confirm.exists()).toBe(true);
  const checkbox = wrapper.find('input[type="checkbox"]');
  expect(checkbox.exists()).toBe(true);
  expect(wrapper.state("purge")).toBe(false);
  checkbox.simulate("change");
  expect(wrapper.state("purge")).toBe(true);

  // Check that the "purge" state is forwarded to deleteApp
  confirm.props().onConfirm(); // Simulate confirmation
  expect(deleteApp).toHaveBeenCalledWith(true);
});

context("when name or namespace do not exist", () => {
  const props = {
    app: new hapi.release.Release({ name: "name", namespace: "my-ns" }),
  };

  itBehavesLike("aLoadingComponent", {
    component: AppControls,
    props: { ...props, app: { ...props.app, name: null } },
  });
  itBehavesLike("aLoadingComponent", {
    component: AppControls,
    props: { ...props, app: { ...props.app, namespace: null } },
  });
});

context("when the application has been already deleted", () => {
  const props = {
    app: new hapi.release.Release({ name: "name", namespace: "my-ns", info: { deleted: {} } }),
    deleteApp: jest.fn(() => false), // Return "false" to avoid redirect when mounting
  };

  it("should show Purge instead of Delete in the button title", () => {
    const wrapper = shallow(<AppControls {...props} />);
    const button = wrapper.find(".button-danger");
    expect(button.text()).toBe("Purge");
  });

  it("should not show the purge checkbox", () => {
    // mount() is necessary to render the Modal
    const wrapper = mount(<AppControls {...props} />);
    ReactModal.setAppElement(document.createElement("div"));
    wrapper.setState({ modalIsOpen: true });
    wrapper.update();

    const confirm = wrapper.find(ConfirmDialog);
    expect(confirm.exists()).toBe(true);
    const checkbox = wrapper.find('input[type="checkbox"]');
    expect(checkbox).not.toExist();
  });

  it("should purge when clicking on delete", () => {
    // mount() is necessary to render the Modal
    const deleteApp = jest.fn(() => false);
    const wrapper = mount(<AppControls {...props} deleteApp={deleteApp} />);
    ReactModal.setAppElement(document.createElement("div"));
    wrapper.setState({ modalIsOpen: true, purge: false });
    wrapper.update();

    const confirm = wrapper.find(ConfirmDialog);
    expect(confirm.exists()).toBe(true);

    // Check that the "purge" is forwarded to deleteApp
    confirm.props().onConfirm(); // Simulate confirmation
    expect(deleteApp).toHaveBeenCalledWith(true);
  });
});
