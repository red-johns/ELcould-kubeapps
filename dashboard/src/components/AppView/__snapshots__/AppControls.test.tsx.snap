// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when name or namespace do not exist loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
/>
`;

exports[`when name or namespace do not exist loading spinner matches the snapshot 2`] = `
<LoadingWrapper
  loaded={false}
/>
`;
